std/machines/hash/keccak.asm

```diff
@@ -0,0 +1,544 @@
+use std::array;
+use std::utils;
+use std::utils::unchanged_until;
+use std::utils::force_bool;
+use std::convert::expr;
+
+machine Keccak with
+    degree: 24,
+    latch: final_step,
+    operation_id: operation_id,
+    call_selectors: sel, // is this needed?
+{
+    // operation keccakf<0> preimage[0], preimage[1], ... -> a_prime_prime_prime[0, 0, 0], a_prime_prime_prime[0, 0, 1], ....
+    col witness operation_id;
+
+    let NUM_ROUNDS: int = 24;
+
+    // pub struct KeccakCols<T> {
+    //     /// The `i`th value is set to 1 if we are in the `i`th round, otherwise 0.
+    //     pub step_flags: [T; NUM_ROUNDS],
+
+
+    //     /// A register which indicates if a row should be exported, i.e. included in a multiset equality
+    //     /// argument. Should be 1 only for certain rows which are final steps, i.e. with
+    //     /// `step_flags[23] = 1`.
+    //     pub export: T,
+
+    //     /// Permutation inputs, stored in y-major order.
+    //     pub preimage: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     pub a: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     /// ```ignore
+    //     /// C[x] = xor(A[x, 0], A[x, 1], A[x, 2], A[x, 3], A[x, 4])
+    //     /// ```
+    //     pub c: [[T; 64]; 5],
+
+    //     /// ```ignore
+    //     /// C'[x, z] = xor(C[x, z], C[x - 1, z], C[x + 1, z - 1])
+    //     /// ```
+    //     pub c_prime: [[T; 64]; 5],
+
+    //     // Note: D is inlined, not stored in the witness.
+    //     /// ```ignore
+    //     /// A'[x, y] = xor(A[x, y], D[x])
+    //     ///          = xor(A[x, y], C[x - 1], ROT(C[x + 1], 1))
+    //     /// ```
+    //     pub a_prime: [[[T; 64]; 5]; 5],
+
+    //     /// ```ignore
+    //     /// A''[x, y] = xor(B[x, y], andn(B[x + 1, y], B[x + 2, y])).
+    //     /// ```
+    //     pub a_prime_prime: [[[T; U64_LIMBS]; 5]; 5],
+
+    //     /// The bits of `A''[0, 0]`.
+    //     pub a_prime_prime_0_0_bits: [T; 64],
+
+    //     /// ```ignore
+    //     /// A'''[0, 0, z] = A''[0, 0, z] ^ RC[k, z]
+    //     /// ```
+    //     pub a_prime_prime_prime_0_0_limbs: [T; U64_LIMBS],
+    // }
+
+    pol commit export;
+    pol commit preimage[100]; // 5 * 5 * 4
```

Comment:
Why not `pol commit preimage[5 * 5 * 4]`?

Reply:
Agree